### PR TITLE
Combined info

### PR DESCRIPTION
**Pull request by JiaruFu**
The condor_log_Integer.py can only deal with selection and groupby of jobs with integer attributes and so far can output the longest-running jobs and times when a certain group of jobs is given.

The following image shows the command format of selection:
Screen Shot 2020-06-03 at 8 56 58 AM

GT means greater than and LT means less than. Other commands include GTE (greater than or equal), LTE (less than or equal), and E (equal). It is case-sensitive. 

This image also contains the groupby command:
Screen Shot 2020-06-03 at 8 56 47 AM

### CODE REVIEW

File: condor_log_Integer.py [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+import getpass
+import itertools
+import sys
+import textwrap
+import enum
+from datetime import datetime, timedelta
+import os
+import time
+import operator
+import graph_job_or_cluster
```
Comment (JoshKarpel) [inline condor_log_Integer.py:12]: This import didn't work when I checked the repository out locally. Perhaps a file that didn't get added to git?

File: condor_log_Integer.py [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+    parser.add_argument(
+        "-debug", action="store_true", help="Turn on HTCondor debug printing."
+    )
+
+    parser.add_argument(
+        "-where",
+        metavar="N",
+        type=str,
+        nargs="+",
+        help="Type in attributes to show certain jobs.",
```
Comment (JoshKarpel) [inline condor_log_Integer.py:51]: This help message is not very informative (same for the `-groupby` help below). The help message should be a short, specific summary of what the option does. A few longer usage examples, like you have in the PR description, written into the program description (in the `ArgumentParser`) would be good as well.

File: condor_log_Integer.py [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+
+    if event_logs is not None:
+        for f in event_logs:
+
+            jel = htcondor.JobEventLog(f)
+            job_time_slot = {}
+            job_event = []
+            returned_job = {}
+
+            new_attributes = {}
```
Comment (JoshKarpel) [inline condor_log_Integer.py:89]: My IDE claims this variable is unused.

File: condor_log_Integer.py [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+            new_conditions = {}
+            false_type_in = []
+            false_type_in_attri = []
+            projected_job = []
+            no_condition = False
+            no_attributes = False
+
+            isIncluded = {}
```
Comment (JoshKarpel) [inline condor_log_Integer.py:97]: Use `snake_case`, not `camelCase`, for variable names.

File: condor_log_Integer.py [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+
+    return condor_log(
+        event_logs=args.files, conditions=args.where, attributes=args.groupby
+    )
+
+
+def condor_log(event_logs=None, conditions=None, attributes=None):
+
+    if event_logs is not None:
+        for f in event_logs:
```
Comment (JoshKarpel) [inline condor_log_Integer.py:82]: It looks like all the analysis is going to be done per log file. We definitely want to be able to ingest multiple log files and do the analysis on all of the jobs in all of them collectively.

File: condor_log_Integer.py [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+            if attributes is None:
+                no_attributes = True
+
+            for event in jel.events(0):
+                job_status = event.type
+
+                if str(job_status) not in job_event:
+                    job_event.append(str(job_status))
+
+                job_c_id = str(event.cluster) + "." + str(event.proc)
```
Comment (JoshKarpel) [inline condor_log_Integer.py:116]: I think this variable is slightly misnamed - `cluster.proc` is the full `job_id`, not the `job_cluster_id`, which I assume is what you're abbreviating to here.

File: condor_log_Integer.py [hunk 7]
```diff
@@ -0,0 +1,655 @@
...
+                if job_c_id not in isIncluded.keys():
+                    isIncluded[job_c_id] = True
+
+                info_dic = {}
+
+                if str(job_status) == JobStatus.JOB_INFO.value:
```
Comment (JoshKarpel) [inline condor_log_Integer.py:122]: This seems too convoluted to me - the job status can change when we get  events, but the event types are not themselves job statuses. This would be simpler as `if event.type == htcondor.JobEventType.JOB_AD_INFORMATION`. The `JobEventType` enumeration is documented [here](https://htcondor.readthedocs.io/en/latest/apis/python-bindings/api/htcondor.html#htcondor.JobEventType).

File: condor_log_Integer.py [hunk 8]
```diff
@@ -0,0 +1,655 @@
...
+                job_status = event.type
+
+                if str(job_status) not in job_event:
+                    job_event.append(str(job_status))
+
+                job_c_id = str(event.cluster) + "." + str(event.proc)
+                if job_c_id not in isIncluded.keys():
+                    isIncluded[job_c_id] = True
+
+                info_dic = {}
```
Comment (JoshKarpel) [inline condor_log_Integer.py:120]: This variable is only used inside the `if` below, so it doesn't need to be pre-declared.

File: condor_log_Integer.py [hunk 9]
```diff
@@ -0,0 +1,655 @@
...
+    if args.debug:
+        print("Enabling HTCondor debug output...")
+        htcondor.enable_debug()
+
+    return condor_log(
+        event_logs=args.files, conditions=args.where, attributes=args.groupby
+    )
+
+
+def condor_log(event_logs=None, conditions=None, attributes=None):
```
Comment (JoshKarpel) [inline condor_log_Integer.py:79]: This function is way too deeply nested for me to follow the logic. I'd encourage you to break it up into smaller functions and use [guard clauses](https://refactoring.guru/replace-nested-conditional-with-guard-clauses) to reduce the amount of nested conditionals.

File: condor_log_Integer.py [hunk 10]
```diff
@@ -0,0 +1,655 @@
...
+    E = "E"
+
+    def __str__(self):
+        return self.value
+
+
+HEADERS = ["CLUSTER"] + ["JOBS"] + list(JobStatus.ordered())
+
+
+class summaries:
```
Comment (JoshKarpel) [inline condor_log_Integer.py:230]: Should be `class Summaries`.

File: condor_log_Integer.py [hunk 11]
```diff
@@ -0,0 +1,655 @@
...
+                    if summ[i] == None:
+                        summ[i] = time[i]
+                    else:
+                        if time[i] != None:
+                            summ[i] = summ[i] + time[i]
+                line = [str(cluster_id)] + [str(job_id)] + time
+
+                # print("".join("{:<15}".format(str(m)) for m in line))
+
+                temp_id = str(cluster_id) + "." + str(job_id)
```
Comment (JoshKarpel) [inline condor_log_Integer.py:263]: This is another area to try to breaking out into functions and apply some guard clauses. Deeply nested logic like this is very hard to follow.

File: condor_log_Integer.py [hunk 12]
```diff
@@ -0,0 +1,655 @@
...
+    def duration(job_time, job_event):
+        event = dict((e, False) for e in job_event)
+        next_event = {}
+        duration = dict(zip(list(JobStatus.ordered()), [None, None, None]))
+        for i in range(len(job_time)):
+            if (i + 1) < len(job_time) and event[job_time[i][0]] == False:
+                if job_time[i][0] == JobStatus.JOB_HELD.value:
+                    j = i
+                    while (
+                        job_time[j][0] not in JobStatus._value2member_map_
```
Comment (JoshKarpel) [inline condor_log_Integer.py:332]: Woah... we definitely shouldn't be digging around in the enum internals like this. Two options: make sure you always use the enum members instead of their values; or if you must mix them, make it a "string enum" (like https://docs.python.org/3/library/enum.html#others, but `str` instead of `int`). Then the enum members really are strings, and they are equal to their values.

File: condor_log_Integer.py [hunk 13]
```diff
@@ -0,0 +1,655 @@
...
+                duration[JobStatus.IDLE] = (
+                    diff
+                    if duration[JobStatus.IDLE] == None
+                    else duration[JobStatus.IDLE] + diff
+                )
+
+        return duration
+
+
+class format_change:
```
Comment (JoshKarpel) [inline condor_log_Integer.py:379]: Again, class names should be like `FormatChange`. Again, since this class has no state, its method should just be a function.

File: condor_log_Integer.py (hunk 10 above)
Comment (JoshKarpel) [reply condor_log_Integer.py:230]: Also, this class has no state and only has static methods. It should be a function.

File: condor_log_Integer.py [hunk 14]
```diff
@@ -0,0 +1,655 @@
...
+                and Signs.GTE.value not in c
+                and Signs.LTE.value not in c
+            ) or wrong_format:
+                if c not in false_type_in:
+                    false_type_in.append(c)
+
+        return new_conditions, false_type_in
+
+
+class info:
```
Comment (JoshKarpel) [inline condor_log_Integer.py:527]: Same as the other classes you have down here: this isn't actually a class.

File: condor_log_Integer.py [hunk 15]
```diff
@@ -0,0 +1,655 @@
...
+                        for all_num in conditions[attri_keys][sign]:
+                            job_id = str(event.cluster) + "." + str(event.proc)
+                            if int(info_dic[attri_keys]) == int(all_num):
+                                if job_id not in projected_job and isIncluded[job_id]:
+                                    projected_job.append(job_id)
+                            else:
+                                if job_id in projected_job:
+                                    projected_job.remove(job_id)
+                                else:
+                                    isIncluded[job_id] = False
```
Comment (JoshKarpel) [inline condor_log_Integer.py:610]: You can replace the switch on the `sign` with a dictionary lookup that returns functions from the `operator` module ([docs](https://docs.python.org/3.8/library/operator.html)).

File: condor_log_Integer.py [hunk 16]
```diff
@@ -0,0 +1,655 @@
...
+        event, attributes, projected_job, info_dic, returned_job, false_type_in_attri
+    ):
+        job_id = str(event.cluster) + "." + str(event.proc)
+        if job_id in projected_job:
+
+            if attributes is not None:
+                attri_keys = info_dic.keys()
+                for at in attributes:
+
+                    if at in attri_keys:
```
Comment (JoshKarpel) [inline condor_log_Integer.py:630]: "attr" is the more common abbreviation of "attribute" in Python code.

File: condor_log_Integer.py [hunk 17]
```diff
@@ -0,0 +1,655 @@
...
+            if conditions is not None:
+                new_conditions, false_type_in = format_change.condition_format(
+                    conditions, false_type_in
+                )
+
+            if conditions is None:
+                no_condition = True
+
+            if attributes is None:
+                no_attributes = True
```
Comment (JoshKarpel) [inline condor_log_Integer.py:108]: It seems like `no_condition/attributes` are just used below a single time each to recheck this `if` condition... I think we don't need the extra indirection of these variables, and can just check whether `conditions/attributes is None` later.

File: condor_log_Integer.py [hunk 18]
```diff
@@ -0,0 +1,655 @@
...
+
+        return duration
+
+
+class format_change:
+    def condition_format(conditions, false_type_in):
+        new_conditions = {}
+        wrong_format = False
+        for c in conditions:
+            if Signs.GT.value in c and Signs.E.value not in c and not wrong_format:
```
Comment (JoshKarpel) [inline condor_log_Integer.py:384]: This whole setup is way too complicated. If the concern is special characters, you can get them to pass through the shell by surrounding them in quotes, e.g. `python condor_log_Integer.py -where 'Cluster > 0'`. This bcomes  a single argument inside Python, the string `"Cluster > 0"`, spaces and all.

File: condor_log_Integer.py [hunk 19]
```diff
@@ -0,0 +1,655 @@
...
+                        summ[i] = time[i]
+                    else:
+                        if time[i] != None:
+                            summ[i] = summ[i] + time[i]
+                line = [str(cluster_id)] + [str(job_id)] + time
+
+                # print("".join("{:<15}".format(str(m)) for m in line))
+
+                temp_id = str(cluster_id) + "." + str(job_id)
+                if type(returned_job) == dict:
```
Comment (JoshKarpel) [inline condor_log_Integer.py:264]: I would recommend normalizing however you produce `returned_job` so that it is always a dictionary. That is, assume that the jobs have been grouped - if there's just one group and it's key is `None`, that means we didn't actually group anything. Then the logic here can hopefully be basically the same whether it's a list or a group.

File: condor_log_Integer.py (hunk 7 above)
Comment (JoshKarpel) [reply condor_log_Integer.py:122]: Also, because you only looks for jobs that have info events, no jobs that don't have info events are ever included in the search. This is surprising - they should act like they have no attributes, not that they don't exist. I recommend using the submit events to record jobs existing, and adding metadata to them when you see a job ad information event for them.

File: condor_log_Integer.py [hunk 20]
```diff
@@ -0,0 +1,655 @@
...
+            n = {lst[0]: lst[1]}
+            info_dic.update(n)
+        return info_dic
+
+    def project(event, conditions, info_dic, projected_job, false_type_in, isIncluded):
+        for attri_keys in conditions.keys():
+
+            if attri_keys in info_dic.keys():
+                values = list(itertools.chain(*conditions[attri_keys].values()))
+                keys = list(conditions[attri_keys].keys())
```
Comment (JoshKarpel) [inline condor_log_Integer.py:549]: My IDE claims these are unused.

File: condor_log_Integer.py [hunk 21]
```diff
@@ -0,0 +1,655 @@
...
+
+                for sign in conditions[attri_keys].keys():
+                    if sign == Signs.GT.value:
+                        for all_num in conditions[attri_keys][sign]:
+                            job_id = str(event.cluster) + "." + str(event.proc)
```
Comment (JoshKarpel) [inline condor_log_Integer.py:554]: You do this construction a lot - make it a function `job_id_from_event(event) -> str`.

File: condor_log_Integer.py [hunk 22]
```diff
@@ -0,0 +1,655 @@
...
+                if cluster_id not in job_time_slot:
+                    job_time_slot[cluster_id] = {}
+
+                job_id = event.proc
+                if job_id not in job_time_slot[cluster_id].keys():
+                    job_time_slot[cluster_id][job_id] = {}
+
+                timestamp = event.timestamp
+                dt_object = datetime.fromtimestamp(timestamp)
+                job_time_slot[cluster_id][job_id][str(job_status)] = dt_object
```
Comment (JoshKarpel) [inline condor_log_Integer.py:158]: I still really want to see a `Job` class that can capture this data in an object.

**Review by JoshKarpel — CHANGES_REQUESTED** [review]
This needs a lot of work, enough that I think we should start from scratch with a simpler model of what's going on. Please read the line-level comments, but also let me know what you think about the plan I propose below.

I propose this as a starting point: write a script that, given some event logs, produces a `Job` object for each job described by those logs. Each `Job` object has a list inside it, `Job.events`, that holds the time-ordered list of events for that job from the event logs. The job's cluster and proc ids are also available as attributes, `Job.cluster` and `Job.proc`. All other information can be derived from the events. This will give us a foundation to build filtering/grouping and statistics/plotting on top of. Organizing the data is the critical first step, so let's get that right before moving on.

File: condor_log_Integer.py (hunk 5 above)
Comment (JiaruFu) [reply condor_log_Integer.py:82]: Does this mean the same job could be on different log files? If it is, does it mean that job could have different submitting times?

Comment (JoshKarpel) [reply condor_log_Integer.py:82]: No, each job will only appear in a single log file. I might want to "combine" multiple log files to generate a larger group of jobs though.
